internal/api/mfa.go

```diff
@@ -61,6 +85,168 @@ const (
+		return nil, badRequestError(ErrorCodeValidationFailed, "WebAuthn RP ID cannot be empty")
+	}
+	if len(config.RPOrigins) == 0 {
+		return nil, badRequestError(ErrorCodeValidationFailed, "WebAuthn RP Origins cannot be empty")
+	}
+
+	var invalidOrigins []string
+
+	for _, origin := range config.RPOrigins {
+		parsedURL, err := url.Parse(origin)
+		if err != nil || (parsedURL.Scheme != "https" && parsedURL.Scheme != "http") || parsedURL.Host == "" {
+			invalidOrigins = append(invalidOrigins, origin)
+		}
+	}
+	if len(invalidOrigins) > 0 {
+		return nil, badRequestError(ErrorCodeValidationFailed, fmt.Sprintf("Invalid RP origins found: %v", invalidOrigins))
+	}
+	wconfig := &webauthn.Config{
+		RPDisplayName: config.RPDisplayName,
+		RPID:          config.RPID,
+		RPOrigins:     config.RPOrigins,
+	}
+	webAuthn, err := webauthn.New(wconfig)
+	if err != nil {
+		return nil, badRequestError(ErrorCodeValidationFailed, fmt.Sprintf("invalid webauthn configuration: %v", err))
+	}
+
+	return webAuthn, nil
+}
+
+func (a *API) handleWebAuthnEnrollment(w http.ResponseWriter, r *http.Request, params *EnrollFactorParams) error {
+	ctx := r.Context()
+	user := getUser(ctx)
+	db := a.db.WithContext(ctx)
+	ipAddress := utilities.GetIPAddress(r)
+
+	webAuthn, err := initializeWebAuthnConfig(params.WebAuthn)
+	if err != nil {
+		return err
+	}
+	options, session, err := webAuthn.BeginRegistration(user)
+	if err != nil {
+		return internalServerError("error generating webauthn registration data").WithInternalError(err)
+	}
+	ws := &models.WebauthnSession{
+		SessionData: session,
+	}
+	factor := models.NewFactor(user, params.FriendlyName, models.WebAuthn, models.FactorStateUnverified)
+	challenge := ws.ToChallenge(factor.ID, ipAddress)
+	err = db.Transaction(func(tx *storage.Connection) error {
+		if terr := tx.Create(factor); err != nil {
+			return terr
+		}
+		if terr := tx.Create(challenge); terr != nil {
+			return terr
+		}
+		return nil
+
+	})
+	if err != nil {
+		return err
+	}
+
+	return sendJSON(w, http.StatusOK, &EnrollWebAuthnFactorResponse{
+		PublicKeyCredentialRequestOptions: options,
+		FactorID:                          factor.ID,
+		ChallengeID:                       challenge.ID,
+		FriendlyName:                      factor.FriendlyName,
+	})
+
+}
+
+func (a *API) handleWebAuthnVerification(w http.ResponseWriter, r *http.Request, params *VerifyFactorParams) error {
+	ctx := r.Context()
+	user := getUser(ctx)
+	config := a.config
+	factor := getFactor(ctx)
+	webAuthn, err := initializeWebAuthnConfig(&params.WebAuthn)
+	if err != nil {
+		return err
+	}
+	db := a.db.WithContext(ctx)
+	challenge, err := models.FindChallengeByID(a.db, params.ChallengeID)
+	if err != nil {
+		return err
+	}
+	sessionData := challenge.ToSession(user.ID, config.MFA.ChallengeExpiryDuration)
+
+	var credential *webauthn.Credential
+
+	if factor.IsUnverified() {
+		credential, err = webAuthn.FinishRegistration(user, sessionData, r)
+		if err != nil {
+			return err
+		}
+
```

Comment:
I don't think you can just use [`FinishRegistration`](https://github.com/go-webauthn/webauthn/blob/v0.11.0/webauthn/registration.go#L207-L214) here. Look at the implementation. It's reading off the request's body.

You probably need to do the internals of that function in this `if` block. Probably for `FinishLogin` below too.